File: src/mocks/lib/paginate.ts
export function paginate<T>(items: T[], page: number, pageSize: number) {
  const safePage = Math.max(1, page)
  const safeSize = Math.max(1, pageSize)
  const start = (safePage - 1) * safeSize
  return {
    items: items.slice(start, start + safeSize),
    total: items.length,
    page: safePage,
    pageSize: safeSize,
  }
}

File: apps/backend/internal/pkg/paginate.go
package pkg

func Paginate[T any](items []T, page, pageSize int) (result []T, total, safePage, safeSize int) {
	safePage = page
	if safePage < 1 {
		safePage = 1
	}
	safeSize = pageSize
	if safeSize < 1 {
		safeSize = 1
	}
	total = len(items)
	start := (safePage - 1) * safeSize
	if start >= total {
		return []T{}, total, safePage, safeSize
	}
	end := start + safeSize
	if end > total {
		end = total
	}
	return items[start:end], total, safePage, safeSize
}
